text/0000-source-date-epoch.md

```diff
@@ -0,0 +1,84 @@
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+  Users have asked for it
+- What use cases does it support?
+  A meaningful image create time, e.g., the ability to determine which images are most recent
+- What is the expected outcome?
+  Images built with `SOURCE_DATE_EPOCH` set will not be reproducible, as the config blob will change
+
+# What it is
+[what-it-is]: #what-it-is
+
+- Define the target persona: buildpack user, platform operator
+  * A buildpack user could pass a flag to `pack` to set `SOURCE_DATE_EPOCH` to the current time in the lifecycle's execution environment during `export`.
+  * A platform operator could choose to set `SOURCE_DATE_EPOCH` whenever `export` is run.
+
+- If applicable, describe the differences between teaching this to existing users and new users.
+  We should mention this feature in the lifecycle release notes and docs.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+This [PR](https://github.com/buildpacks/imgutil/pull/137) to imgutil would read the environment variable at the point of saving the image. An alternative implementation could be to have the `exporter` read the variable and provide it via a new `SetCreatedAt()` method on the `imgutil.Image` interface. The latter might be safer as it would avoid unintended side effects for other consumers of imgutil (e.g., `pack`). 
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+It breaks build reproducibility. However platforms must explicitly opt-in to this feature.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+  * Doing nothing
+  * Allowing layer timestamps to also be configurable, but this would effectively prohibit launch layer re-use and make builds slower
+- Why is this proposal the best?
+  * It is easy to implement, easy to use, and solves for the desired use case
+- What is the impact of not doing this?
+  * The lack of meaningful image create times makes it difficult to automate cleanup of images and could be a deal breaker for some users
+
+# Prior Art
+[prior-art]: #prior-art
+
+* See under "Reading the variable" [here](https://reproducible-builds.org/docs/source-date-epoch/)
+* [ko](https://github.com/google/ko#why-are-my-images-all-created-in-1970)
+* [jib](https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#why-is-my-image-created-48-years-ago)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- What parts of the design do you expect to be resolved before this gets merged?
+  * Should we require users to be on platform api 0.9 in order to use this feature? (Opinion: no, because this feature is invisible if the environment variable is not set. The motivation for spec'ing this in the platform api is to make it clear for platform operators how to use it.)
```

Comment:
I agree - no need to be on newer platforms to use it.

Reply:
If we are not tying it to a specific platform API, I don't see why we should note this in the platform spec and not in our docs. It seems a weird that we are retroactively adding something to the past platform APIs but only documenting it for 0.9. I would be in favor of just documenting this in the lifecycle docs somewhere and implementing it in pack via a flag if we plan on introducing this for past platform APIs that have already been released.

Reply:
We follow something similar with our estargz support in lifecycle.

Reply:
I disagree that this shouldn't be part of the specification. As platform implementors, we rely on the API to know what parameters we can use especially when it has specific side effects on the outcome. `estargz` support is an accident and can't be documented as supported because of the similar.

Essentially platforms that don't get to select their lifecycle implementations but rely on what's on the builder go from saying _"by doing X you get Y"_ to _"by doing X you MAY get Y"_.

Reply:
To the original opinion regarding backward compatibility, nothing on the specification says a lifecycle implementation _CAN'T_ support an arbitrary environment variable.

Reply:
Yes, but it very much violates the whole point of having a platform API. If we start introducing new features for older platform APIs, the platform is now dependent on a specific lifecycle version to support a specific feature rather than knowing the feature set it will have available given a platform API.

Say you are a platform implementing this feature in an older platform api and you set this environment variable. You as a platform will never know whether or not the lifecycle actually respected the variable and set an appropriate date for the image or not as the lifecycle has no such contract for older apis.

If we implement this as part of a new platform api then there is an inherent contract that the platform can rely upon.

Reply:
> the lifecycle has no such contract for older apis

Personally I think this is totally fine. The feature isn't mentioned in older platform apis, so it can't be relied upon. Use at your own risk.

Reply:
I think we should spec the feature b/c it's existence determines whether a platform like `pack` can provide a new feature (a flag to use the real creation time) to users.

> the platform is now dependent on a specific lifecycle version to support a specific feature rather than knowing the feature set it will have available given a platform API.

Whatever decision we make here, I still think the platform API is the source of truth.

With my platform author hat on, I basically see the decision to support this in old APIs as irrelevant. I won't trust that I can use it in an old API, but also it doesn't hurt me. Maybe there are some edge cases where having it is helpful. E.g. I operate a very specific platform that uses a specific builder and I know my lifecycle version, maybe I go ahead and use this feature without checking the lifecycle version? But I am only tempted if I control the environment and know what I am doing.

That was a very long winded way of saying "whatever y'all can agree on is fine with me on this one".

Reply:
:+1: on it being spec'd if we want it as part of the contract.